clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -23,239 +23,410 @@ package common
+		isChecking = atomic.LoadInt32(&w.runningRoutines) > 0
+		// Check if events versions caught up with listed objects versions.
+		isUpToDate = w.opResourceVersion >= maxResourceVersion
+		numberRunning = 0
+		for _, checker := range w.checkerMap {
+			if status := checker.getStatus(); status {
+				numberRunning++
 			}
+		}
+		w.lock.Unlock()
+		if !isChecking && isUpToDate && desiredCount == numberRunning {
+			break
+		}
+	}
 
-			atomic.AddInt32(&runningRuntimeObjects, 1)
-			objName, err := getNameFromRuntimeObject(runtimeObjectsList[objectIndex])
-			if err != nil {
-				errList.Append(fmt.Errorf("reading object name error: %v", err))
-				return
-			}
-			objNamespace, err := getNamespaceFromRuntimeObject(runtimeObjectsList[objectIndex])
-			if err != nil {
-				errList.Append(fmt.Errorf("reading object namespace error: %v", err))
-				return
-			}
-			glog.Infof("%s: %s (%s) has all pods running", waitForControlledPodsRunningName, objName, objNamespace)
-		})
+	if isChecking || !isUpToDate || desiredCount != numberRunning {
+		return fmt.Errorf("timed out while waiting for controlled pods")
 	}
-	wg.Wait()
+	glog.Infof("%s: %d/%d %ss are running with all pods", waitForControlledPodsRunningName, numberRunning, desiredCount, w.kind)
+	return nil
+}
 
-	if int(runningRuntimeObjects) == len(runtimeObjectsList) {
-		glog.Infof("%s: %d / %d %ss are running with all pods", waitForControlledPodsRunningName, runningRuntimeObjects, len(runtimeObjectsList), kind)
-		return summaries, nil
+func (w *waitForControlledPodsRunningMeasurement) stop() {
+	close(w.stopCh)
+	w.queue.ShutDown()
+	w.wg.Wait()
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.isRunning = false
+	for _, checker := range w.checkerMap {
+		checker.terminate()
 	}
-	return summaries, fmt.Errorf("error while waiting for %ss to be running in namespace '%v' with labels '%v' and fields '%v' - only %d found running with all pods: %v",
-		kind, namespace, labelSelector, fieldSelector, runningRuntimeObjects, errList.String())
 }
 
-func waitForRuntimeObject(c clientset.Interface, obj runtime.Object, timeout time.Duration) error {
-	runtimeObjectNamespace, err := getNamespaceFromRuntimeObject(obj)
+func (w *waitForControlledPodsRunningMeasurement) handleObject(c clientset.Interface, obj runtime.Object) {
+	objName, err := runtimeobjects.GetNameFromRuntimeObject(obj)
 	if err != nil {
-		return err
+		glog.Errorf("%s: reading object name error: %v", waitForControlledPodsRunningName, err)
+		return
 	}
-	runtimeObjectSelector, err := getSelectorFromRuntimeObject(obj)
+	objNamespace, err := runtimeobjects.GetNamespaceFromRuntimeObject(obj)
 	if err != nil {
-		return err
+		glog.Errorf("%s: reading object namespace error: %v", waitForControlledPodsRunningName, err)
+		return
 	}
-	runtimeObjectReplicas, err := getReplicasFromRuntimeObject(obj)
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	if !w.isRunning {
+		return
+	}
+	key := createMetaNamespaceKey(objNamespace, objName)
```

Comment:
Can you make this function to take runtime.Object instead?

The longer function you write, the harder it is to follow.

Reply:
Done.